File: tests/xcalablemp/global-view/template_fix/F/reflect.f90
program tfix

  integer m(4)
  integer, parameter :: N = 120
  integer xmp_node_num
  integer im1

!$xmp nodes p(4)

!$xmp template t(:)
!$xmp distribute t(gblock(*)) onto p

  integer, allocatable ::  a(:)
!$xmp align a(i) with t(i)
!$xmp shadow a(2)

  m = (/ 8, 16, 32, 64 /)

!$xmp template_fix (gblock(m)) t(N)

  allocate (a(N))

!$xmp loop (i) on t(i)
  do i = 1, N
     a(i) = i
  end do

!$xmp reflect (a) width (/periodic/1:1) async (100)
!$xmp wait_async (100)

!$xmp loop (i) on t(i)
  do i = 1, N
     im1 = i - 1
     if (a(i-1) /= mod(im1-1+N, N)+1) then
        write(*,*) "ERROR: Lower in", i, a(i-1), mod(im1-1+N, N)+1
        call exit(1)
     end if
     if (a(i+1) /= mod(im1+1+N, N)+1) then
        write(*,*) "ERROR: Upper in", i, a(i+1), mod(im1+1+N, N)+1
        call exit(1)
   end if
end do

!$xmp task on p(1)
  write(*,*) "PASS"
!$xmp end task

end program tfix

File: tests/xcalablemp/global-view/template_fix/C/reflect.c
#include <stdio.h>
#include <stdlib.h>     
#include "xmp.h"

#define N 120
int m[4] = { 8, 16, 32, 64 };
int NSIZE = N;

#pragma xmp nodes p(4)
#pragma xmp template t(:)
#pragma xmp distribute t(gblock(*)) onto p

int xmp_node_num();
void foo()
{
#pragma xmp template_fix (gblock(m)) t(0:NSIZE-1)
  int *a;
#pragma xmp align a[i] with t(i)
#pragma xmp shadow a[2]

  a = (int*)xmp_malloc(xmp_desc_of(a), NSIZE);

#pragma xmp loop on t(i)
  for (int i = 0; i < N; i++)
    a[i] = i;

#pragma xmp reflect (a) width (/periodic/1:1) async (100)
#pragma xmp wait_async (100)

#pragma xmp loop (i) on t(i)
  for (int i = 0; i < N; i++){
    if(a[i-1] != (i-1+N)%N){
      printf("ERROR Lower in %d (%06d) (%06d)\n", (int)i, a[i-1], (int)(i-1+N)%N);
      exit(1);
    }
    if(a[i+1] != (i+1+N)%N){
      printf("ERROR Upper in %d (%06d) (%06d)\n", (int)i, a[i+1], (int)(i+1+N)%N);
      exit(1);
    }
  }

#pragma xmp task on p(1)
  {
    printf("PASS\n");
  }
}

int main()
{
  foo();
  return 0;
}
